data/transactions/logic/resources.go

```diff
@@ -223,24 +235,30 @@ func (cx *EvalContext) requireHolding(acct basics.Address, id basics.AssetIndex)
 		return nil
 	}
 	if !cx.allowsHolding(acct, id) {
-		return fmt.Errorf("unavailable Holding %s x %d would be accessible", acct, id)
+		return fmt.Errorf("unavailable Holding %d+%s would be accessible", id, acct)
 	}
 	return nil
 }
 
 func (cx *EvalContext) requireLocals(acct basics.Address, id basics.AppIndex) error {
 	if !cx.allowsLocals(acct, id) {
-		return fmt.Errorf("unavailable Local State %s x %d would be accessible", acct, id)
+		return fmt.Errorf("unavailable Local State %d+%s would be accessible", id, acct)
 	}
 	return nil
 }
 
 func (cx *EvalContext) allowsAssetTransfer(hdr *transactions.Header, tx *transactions.AssetTransferTxnFields) error {
-	err := cx.requireHolding(hdr.Sender, tx.XferAsset)
-	if err != nil {
-		return fmt.Errorf("axfer Sender: %w", err)
+	// After EnableInnerClawbackWithoutSenderHolding appears in a consensus
+	// update, we should remove it from consensus params and assume it's true in
+	// the next release. It only needs to be in there so that it gates the
+	// beahvior change in the release it first appears.
+	if !cx.Proto.EnableInnerClawbackWithoutSenderHolding || tx.AssetSender.IsZero() {
```

Comment:
are we allowing inner txn to transfer any asset?

Reply:
Inners can already perform clawbacks (their app address must be the clawback address for the asset).

This flag means that when an inner performs such a clawback, their _holding_ is not required to be available.  Their balance in the asset (or even whether they are opted in) is irrelevant, and the transaction code never looks it up. So we should not require it to be available.